sdks/python/apache_beam/io/restriction_trackers.py

```diff
@@ -0,0 +1,117 @@
+
+  def __init__(self, start, stop):
+    if start > stop:
+      raise ValueError(
+          'Start offset must be not be larger than the stop offset. '
+          'Received %d and %d respectively.', start, stop)
+    self.start = start
+    self.stop = stop
+
+  def __eq__(self, other):
+    if not isinstance(other, OffsetRange):
+      return False
+
+    return self.start == other.start and self.stop == other.stop
+
+  def __ne__(self, other):
+    if not isinstance(other, OffsetRange):
+      return True
+
+    return not (self.start == other.start and self.stop == other.stop)
+
+  def split(self, desired_num_offsets_per_split, min_num_offsets_per_split=1):
+    current_split_start = self.start
+    max_split_size = max(desired_num_offsets_per_split,
+                         min_num_offsets_per_split)
+    while current_split_start < self.stop:
+      current_split_stop = min(current_split_start + max_split_size, self.stop)
+      remaining = self.stop - current_split_stop
+
+      # Avoiding a small split at the end.
+      if (remaining < desired_num_offsets_per_split / 4 or
+          remaining < min_num_offsets_per_split):
+        current_split_stop = self.stop
+
+      yield OffsetRange(current_split_start, current_split_stop)
+      current_split_start = current_split_stop
+
+  def new_tracker(self):
+    return OffsetRangeTracker(self.start, self.stop)
+
+
+class OffsetRestrictionTracker(RestrictionTracker):
+  """An `iobase.RestrictionTracker` implementations for byte offsets."""
+
+  def __init__(self, start_position, stop_position):
+    self._range = OffsetRange(start_position, stop_position)
+    self._current_position = None
+    self._last_claim_attempt = None
+    self._checkpointed = False
+    self._lock = threading.Lock()
+
+  def check_done(self):
+    if self._last_claim_attempt < self._range.stop - 1:
+      raise ValueError(
+          'OffsetRestrictionTracker is not done since work in range [%s, %s) '
+          'has not been claimed.',
+          self._last_claim_attempt if self._last_claim_attempt is not None
+          else self._range.start,
+          self._range.stop)
+
+  def current_restriction(self):
+    return (self._range.start, self._range.stop)
+
+  def start_position(self):
+    return self._range.start
+
+  def stop_position(self):
+    return self._range.stop
+
+  def try_claim(self, position):
+    with self._lock:
+      self._last_claim_attempt = position
+      if position >= self._range.start and position < self._range.stop:
+        self._current_position = position
+        return True
+
+      return False
+
+  def checkpoint(self):
+    with self._lock:
+      residual_range = (
+          (self._range.start, self._range.stop)
+          if self._current_position is None
+          else (self._current_position + 1, self._range.stop))
+      # If self._current_position is 'None' no records have been claimed so
+      # residual should start from self._range.start.
```

Comment:
<!--new_thread; commit:7549b5c2ebe2ae47af9066eaf97364a27e828ab5; resolved:0-->
Move this comment to before the if above.